src/apps/rigoblock/common/rigoblock.pool.contract-position-fetcher.ts

```diff
@@ -0,0 +1,144 @@
+  address: string;
+  label: string;
+  underlyingTokenBalances: ContractPositionBalance<UniswapV3LiquidityPositionDataProps>[];
+};
+
+export abstract class RigoblockPoolContractPositionFetcher extends ContractPositionTemplatePositionFetcher<
+  SmartPool,
+  RigoblockLiquidityDataProps,
+  RigoblockPoolAppTokenDefinition
+> {
+  abstract positionManagerAddress: string;
+  abstract groupLabel: string;
+
+  constructor(
+    @Inject(APP_TOOLKIT) protected readonly appToolkit: IAppToolkit,
+    @Inject(RigoblockContractFactory) protected readonly rigoblockContractFactory: RigoblockContractFactory,
+    @Inject(UniswapV3ContractFactory) protected readonly contractFactory: UniswapV3ContractFactory,
+    @Inject(UniswapV3LiquidityContractPositionBuilder)
+    protected readonly uniswapV3LiquidityContractPositionBuilder: UniswapV3LiquidityContractPositionBuilder,
+  ) {
+    super(appToolkit);
+  }
+
+  getContract(address: string): SmartPool {
+    return this.rigoblockContractFactory.smartPool({ address, network: this.network });
+  }
+
+  async getDefinitions(): Promise<RigoblockPoolAppTokenDefinition[]> {
+    const appTokens = await this.appToolkit.getAppTokenPositions({
+      appId: this.appId,
+      network: this.network,
+      groupIds: ['pool'],
+    });
+
+    return await Promise.all(
+      appTokens.map(async pool => {
+        const underlyingTokenBalances = await this.getUnderlyingBalances(pool.address);
+        return {
+          address: pool.address,
+          label: pool.displayProps.label,
+          underlyingTokenBalances: underlyingTokenBalances,
+        };
+      }),
+    );
+  }
+
+  // we define the liquidity position of a rigoblock pool
+  async getTokenDefinitions({ definition }) {
+    return definition.underlyingTokenBalances.map(balance => {
+      return balance.tokens.map(token => {
+        return [
+          {
+            metaType: MetaType.SUPPLIED,
+            address: token.address,
+            network: this.network,
+          },
+        ];
+      });
+    }).flat(2);
+  }
+
+  async getDataProps({ definition }): Promise<RigoblockLiquidityDataProps> {
+    const liquidityBalances = definition.underlyingTokenBalances;
+    const liquidityPositions: UnderlyingLiquidityPositionTokens[] = liquidityBalances.map(balance => {
+      return balance.tokens.map(token => {
+        return {
+          address: token.address,
+          balance: token.balance,
+          balanceRaw: token.balanceRaw,
+          balanceUSD: token.balanceUSD,
+        };
+      });
+    }).flat(1);
+    const liquidity = sumBy(liquidityPositions, v => v.balanceUSD);
+    return {
+      liquidity: liquidity,
+      liquidityPositions: liquidityPositions
+    };
+  }
+
+  async getLabel({ definition }) {
+    return definition.label;
+  }
+
+  async getTokenBalancesPerPosition({ address, contractPosition }) {
+    const { dataProps } = contractPosition;
+    return dataProps.liquidityPositions.map(v => v.balanceRaw);
+  }
```

Comment:
It should fetch balances for a given address

Reply:
I could do the onchain call again, but it's a custom contract, so would have to call this.getUnderlyingBalances again. By initializing the balances in the definitions, the rpc calls can be saved here.